phoenix-core/src/test/java/org/apache/phoenix/index/PrepareIndexMutationsForRebuildTest.java

```diff
@@ -0,0 +1,626 @@
+    }
+
+    @Test
+    public void testCoveredIndexColumns() throws Exception {
+        SetupInfo info = setup(TABLE_NAME,
+                INDEX_NAME,
+                "ROW_KEY VARCHAR, C1 VARCHAR, C2 VARCHAR",
+                "C1",
+                "ROW_KEY",
+                "C2");
+
+        Put dataPut = new Put(Bytes.toBytes(ROW_KEY));
+        addCellToPutMutation(dataPut,
+                info.indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                Bytes.toBytes("C1"),
+                1,
+                Bytes.toBytes("v1"));
+        addCellToPutMutation(dataPut,
+                info.indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                Bytes.toBytes("C2"),
+                1,
+                Bytes.toBytes("v2"));
+        addEmptyColumnToDataPutMutation(dataPut, info.pDataTable, 1);
+
+        Delete dataDel = new Delete(Bytes.toBytes(ROW_KEY));
+        addCellToDelMutation(dataDel,
+                info.indexMaintainer.getEmptyKeyValueFamily().copyBytesIfNecessary(),
+                Bytes.toBytes("C1"),
+                2,
+                KeyValue.Type.DeleteColumn);
+
+        List<Mutation> actualIndexMutations = IndexRebuildRegionScanner.prepareIndexMutationsForRebuild(info.indexMaintainer,
+                dataPut,
+                dataDel);
+
+        List<Mutation> expectedIndexMutations = new ArrayList<>();
+        byte[] idxKeyBytes = generateIndexRowKey("v1");
+
+        Put idxPut1 = new Put(idxKeyBytes);
+        addCellToPutMutation(idxPut1,
+                QueryConstants.DEFAULT_COLUMN_FAMILY_BYTES,
+                Bytes.toBytes("0:C2"),
+                1,
+                Bytes.toBytes("v2"));
+        addEmptyColumnToIndexPutMutation(idxPut1, info.indexMaintainer, 1);
+        expectedIndexMutations.add(idxPut1);
+
+        List<Byte> idxKey2 = new ArrayList<>();
+        idxKey2.add(QueryConstants.SEPARATOR_BYTE);
+        idxKey2.addAll(com.google.common.primitives.Bytes.asList(Bytes.toBytes(ROW_KEY)));
+        byte[] idxKeyBytes2 = com.google.common.primitives.Bytes.toArray(idxKey2);
+        Put idxPut2 = new Put(idxKeyBytes2);
+        addCellToPutMutation(idxPut2,
+                QueryConstants.DEFAULT_COLUMN_FAMILY_BYTES,
+                Bytes.toBytes("0:C2"),
+                2,
+                Bytes.toBytes("v2"));
+        addEmptyColumnToIndexPutMutation(idxPut2, info.indexMaintainer, 2);
+        expectedIndexMutations.add(idxPut2);
+
+        Delete idxDel = new Delete(idxKeyBytes);
+        addCellToDelMutation(idxDel,
+                QueryConstants.DEFAULT_COLUMN_FAMILY_BYTES,
+                null,
+                2,
+                KeyValue.Type.DeleteFamily);
+        expectedIndexMutations.add(idxDel);
+
+        assertEqualMutationList(expectedIndexMutations, actualIndexMutations);
+    }
+
+    @Test
+    public void testForMultipleFamilies() throws Exception {
```

Comment:
testDeleteForMultipleFamilies

Reply:
Actually it tests both put and delete for multiple families case